src/encrypted_container.rs

```diff
@@ -0,0 +1,248 @@
+            }
+            Error::DeserializationToObjectError => {
+                trussed::error::Error::InvalidSerializationFormat
+            }
+            Error::ObjectSerializationError => trussed::error::Error::InvalidSerializationFormat,
+            Error::ContainerSerializationError => trussed::error::Error::InvalidSerializationFormat,
+            Error::SerializationBufferTooSmall => trussed::error::Error::InternalError,
+            Error::FailedEncryption => trussed::error::Error::InternalError,
+            Error::FailedContainerSerialization => {
+                trussed::error::Error::InvalidSerializationFormat
+            }
+            Error::EmptyContainerData => trussed::error::Error::WrongMessageLength,
+            Error::FailedDecryption => trussed::error::Error::InvalidSerializationFormat,
+            Error::EmptyDecryptedData => trussed::error::Error::WrongMessageLength,
+        }
+    }
+}
+
+/// Universal AEAD encrypted data container, using CBOR and Chacha8Poly1305
+///
+/// Encryption is realized by serializing the object using CBOR, then encrypting it using Chacha8Poly1305,
+/// storing related crypto data, namely nonce and tag, and finally serializing the latter,
+/// again using CBOR.
+///
+/// For the plaintext of size 48 bytes, the resulting container size is 87 bytes,
+/// including the 28 bytes of cryptographic data overhead, and leaving 11 bytes
+/// as the CBOR serialization overhead.
+///
+/// Decryption operation is done the same way as its counterpart, but backwards.
+/// The serialized Encrypted Data Container in bytes is first deserialized, making a EDC instance,
+/// and afterwards the decryption operation in Trussed is called, resulting in a original serialized
+/// object, which is then deserialized to a proper instance.
+///
+/// CBOR was chosen as the serialization format due to its simplicity and extensibility.
+/// If that is a requirement, more space efficient and faster would be postcard. Be advised however,
+/// that it's format changes between major revisions (as expected with semver versioning).
+///
+/// This type has implemented bidirectional serialization to trussed Message object.
+///
+/// Showing the processing paths graphically:
+/// T -> \[u8\]: object -> CBOR serialization -> EncryptedDataContainer encryption  -> CBOR serialization -> serialized EncryptedDataContainer
+///
+/// \[u8\] -> T: serialized EncryptedDataContainer -> CBOR deserialization -> EncryptedDataContainer decryption -> CBOR deserialization -> object
+///
+/// Note: to decrease the CBOR overhead it might be useful to rename the serialized object fields for
+/// the serialization purposes. Use the `#[serde(rename = "A")]` attribute.
+///
+/// Usage example:
+/// ```compile_fail
+/// let ek = get_encryption_key(trussed)?;
+/// let data = EncryptedDataContainer::from_obj(trussed, obj, None, ek)?;
+/// let data_serialized: Message = data.try_into()?;
+/// ```
```

Comment:
I try to avoid `compile_fail` doc tests (except to demonstrate compilation errors ;)) because they are very likely to become outdated at some point.  In this case, it looks like `get_encryption_key` is the main problem.  There are two ways to fix that:

1. You can add a commented code that will not show up in the documentation but will make the snippet compile:

```
/// # fn get_encryption_key(trussed: &mut trussed::Client) -> Key { unimplemented!() }
/// let ek = get_encryption_key(trussed)?;
/// let data = EncryptedDataContainer::from_obj(trussed, obj, None, ek)?;
/// let data_serialized: Message = data.try_into()?;
```

2. You can write the example as a method (again using comments to make the displayed snippet more concise):

```
/// # use trussed::Client;
/// # use serde::Serialize;
/// fn encrypt_unit<O: Serialize>(trussed: &mut Client, obj: &O, ek: Key) -> Message {
///    let data = EncryptedDataContainer::from_obj(trussed, obj, None, ek)?;
///    let data_serialized: Message = data.try_into()?;
///    data_serialized
//// }
```